### PR TITLE
GEODE-6900: Add a unit test with detect read conflicts

### PR DESCRIPTION
**Pull request by pivotal-eshu**

### CODE REVIEW

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java [hunk 1]
```diff
@@ -69,7 +81,7 @@ protected void createCachePR() {
     props.put(MCAST_PORT, "0");
```
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java]: These uses of `Properties.put` are pre-existing but should be changed to `Properties.setProperty` if you're up for it.

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java [hunk 2]
```diff
@@ -81,25 +93,95 @@ public void tearDown() throws Exception {
...
+    region.put(key1, value1);
+    TXManagerImpl txManager = (TXManagerImpl) cache.getCacheTransactionManager();
+    txManager.begin();
+    assertThat(region.get(key)).isSameAs(value);
+    region.put(key1, newValue1);
+    TXState txState =
+        (TXState) ((TXStateProxyImpl) TXManagerImpl.getCurrentTXState()).getRealDeal(null, null);
+    txState.setAfterReservation(() -> readTransactionAfterReservation());
+    Runnable task = () -> doPutOnReadKeyTransaction();
+    new Thread(task).start();
```
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java]: I would add an `ExecutorServiceRule` and use that instead. It'll ensure cleanup if the thread gets stuck for any reason and will perform error detection on it.
```
@Rule
public ExecutorServiceRule executorServiceRule = new ExecutorServiceRule();

ExecutorServiceRule.ThrowingRunnable task = () -> { Thread.sleep(100);};
executorServiceRule.submit(task);
```
I'll file a Jira to add support back in to ExecutorServiceRule for java.lang.Runnable as well. ThrowingRunnable allows the body of the run to throw anything.

Comment (pivotal-eshu) [reply geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java]: Thanks for reviewing. Will make the necessary changes.

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java [hunk 3]
```diff
@@ -78,28 +94,94 @@ public void tearDown() throws Exception {
...
+    createCache();
+
+    region.put(key, value);
+    region.put(key1, value1);
+    TXManagerImpl txManager = (TXManagerImpl) cache.getCacheTransactionManager();
+    txManager.begin();
+    assertThat(region.get(key)).isSameAs(value);
+    region.put(key1, newValue1);
+    Runnable task = () -> doPutTransaction();
+    new Thread(task).start();
```
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java]: This should change to:
```
executorServiceRule.submit(() -> doPutTransaction());
```

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java [hunk 4]
```diff
@@ -78,28 +94,94 @@ public void tearDown() throws Exception {
...
+    executorServiceRule.submit(task);
+    txManager.commit();
+    assertThat(region.get(key)).isSameAs(value);
+    assertThat(region.get(key1)).isSameAs(newValue1);
+  }
+
+  private void readTransactionAfterReservation() {
+    allowWriteTransactionToCommitLatch.countDown();
+    try {
+      allowReadTransactionToProceedLatch.await(10, TimeUnit.SECONDS);
```
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java]: I know it seems extreme, but I would just use the `GeodeAwaitility.getTimeout().getValueInMS()`. The reason is that it'll be robust in the face of a long stop-the-world GC or a CI environment with very slow or overloaded CPU:
```
allowReadTransactionToProceedLatch.await(GeodeAwaitility.getTimeout().getValueInMS(), TimeUnit.MILLISECONDS);
```

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java [hunk 5]
```diff
@@ -78,28 +94,94 @@ public void tearDown() throws Exception {
...
+    region.put(key, value);
+    region.put(key1, value1);
+    TXManagerImpl txManager = (TXManagerImpl) cache.getCacheTransactionManager();
+    txManager.begin();
+    assertThat(region.get(key)).isSameAs(value);
+    region.put(key1, newValue1);
+    TXState txState =
+        (TXState) ((TXStateProxyImpl) TXManagerImpl.getCurrentTXState()).getRealDeal(null, null);
+    txState.setAfterReservation(() -> readTransactionAfterReservation());
+    ExecutorServiceRule.ThrowingRunnable task = () -> doPutOnReadKeyTransaction();
```
Comment (kirklund) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java]: You can just inline this without the local variable `task`:
```
executorServiceRule.submit(() -> doPutOnReadKeyTransaction());
```

**Pull request closed without merging**